File: src/devtools/builtin_scripts.py
from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path

from devtools.models import ScriptContext


def run_node_audit_fix(context: ScriptContext) -> dict[str, object]:
    project_root = context.project.path.resolve()
    npm_path = shutil.which("npm")
    if npm_path is None:
        return {"success": False, "message": "npm was not found on PATH."}

    package_dirs = find_project_file_dirs(project_root, "package.json")
    if not package_dirs:
        return {"success": False, "message": "No package.json files found below the selected project."}

    registry = str(context.args.get("registry", "https://registry.npmjs.org"))
    force = bool(context.args.get("force", False))
    command = [npm_path, "audit", "fix"]
    if force:
        command.append("--force")
    command.append(f"--registry={registry}")
    return _run_command_across_dirs(
        project_root=project_root,
        target_dirs=package_dirs,
        command=command,
        success_label="npm audit fix",
    )


def run_maven_dependency_update(context: ScriptContext) -> dict[str, object]:
    project_root = context.project.path.resolve()
    mvn_path = shutil.which("mvn")
    if mvn_path is None:
        return {"success": False, "message": "mvn was not found on PATH."}

    pom_dirs = find_project_file_dirs(project_root, "pom.xml")
    if not pom_dirs:
        return {"success": False, "message": "No pom.xml files found below the selected project."}

    allow_major_updates = bool(context.args.get("allow_major_updates", False))
    command = [
        mvn_path,
        "versions:use-latest-releases",
        "-DgenerateBackupPoms=false",
    ]
    if not allow_major_updates:
        command.append("-DallowMajorUpdates=false")
    command.extend(["-f", "pom.xml"])
    return _run_command_across_dirs(
        project_root=project_root,
        target_dirs=pom_dirs,
        command=command,
        success_label="maven dependency update",
    )


def find_package_dirs(project_root: Path) -> list[Path]:
    return find_project_file_dirs(project_root, "package.json")


def find_project_file_dirs(project_root: Path, filename: str) -> list[Path]:
    discovered: list[Path] = []
    for current_root, dirnames, filenames in os.walk(project_root):
        current_path = Path(current_root)
        dirnames[:] = [dirname for dirname in dirnames if dirname not in {"node_modules", ".git", "dist", "target"}]
        if filename in filenames:
            discovered.append(current_path.resolve())
    return sorted(discovered, key=lambda item: item.as_posix())


def _run_command_across_dirs(
    project_root: Path,
    target_dirs: list[Path],
    command: list[str],
    success_label: str,
) -> dict[str, object]:
    failures: list[str] = []
    for target_dir in target_dirs:
        relative_dir = target_dir.relative_to(project_root)
        label = "." if str(relative_dir) == "." else relative_dir.as_posix()
        print(f"Running in {label}: {' '.join(command)}")

        result = subprocess.run(
            command,
            cwd=target_dir,
            capture_output=True,
            text=True,
            check=False,
        )

        if result.stdout.strip():
            print(result.stdout.rstrip())
        if result.stderr.strip():
            print(result.stderr.rstrip())
        if result.returncode != 0:
            failures.append(label)
            print(f"Command failed in {label} with exit code {result.returncode}")

    if failures:
        summary = ", ".join(failures)
        return {
            "success": False,
            "message": f"{success_label} failed in {len(failures)} location(s): {summary}",
        }
    return {
        "success": True,
        "message": f"{success_label} completed in {len(target_dirs)} location(s).",
    }

File: src/builtin-scripts.ts
import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { spawnSync } from "node:child_process";

import type { BuiltinScriptResponse, ScriptContext } from "./models.ts";

const RUNTIME_SKIP_DIRS = new Set(["node_modules", ".git", "dist", "target", ".angular", ".cache", ".next"]);

export function createRunId(): string {
  return randomUUID().replaceAll("-", "");
}

export function runNodeAuditFix(context: ScriptContext): BuiltinScriptResponse {
  const npmPath = resolveExecutable("npm");
  if (!npmPath) {
    return { success: false, message: "npm was not found on PATH." };
  }

  const projectRoot = path.resolve(context.project.path);
  const packageDirs = findProjectFileDirs(projectRoot, "package.json");
  if (packageDirs.length === 0) {
    return { success: false, message: "No package.json files found below the selected project." };
  }

  const registry = String(context.args.registry ?? "https://registry.npmjs.org");
  const force = Boolean(context.args.force ?? false);
  const command = [npmPath, "audit", "fix"];
  if (force) {
    command.push("--force");
  }
  command.push(`--registry=${registry}`);
  return runCommandAcrossDirs(projectRoot, packageDirs, command, "npm audit fix");
}

export function runMavenDependencyUpdate(context: ScriptContext): BuiltinScriptResponse {
  const mvnPath = resolveExecutable("mvn");
  if (!mvnPath) {
    return { success: false, message: "mvn was not found on PATH." };
  }

  const projectRoot = path.resolve(context.project.path);
  const pomDirs = findProjectFileDirs(projectRoot, "pom.xml");
  if (pomDirs.length === 0) {
    return { success: false, message: "No pom.xml files found below the selected project." };
  }

  const allowMajorUpdates = Boolean(context.args.allow_major_updates ?? false);
  const command = [mvnPath, "versions:use-latest-releases", "-DgenerateBackupPoms=false"];
  if (!allowMajorUpdates) {
    command.push("-DallowMajorUpdates=false");
  }
  command.push("-f", "pom.xml");
  return runCommandAcrossDirs(projectRoot, pomDirs, command, "maven dependency update");
}

export function runEchoProject(context: ScriptContext): BuiltinScriptResponse {
  process.stdout.write(`project=${context.project.name}\n`);
  process.stdout.write(`path=${context.project.path}\n`);
  process.stdout.write(`type=${context.project.projectType}\n`);
  if (Boolean(context.args.include_marker ?? true)) {
    process.stdout.write(`marker=${context.project.marker}\n`);
  }
  return { success: true, message: "Project info printed." };
}

export function runListDirectChildren(context: ScriptContext): BuiltinScriptResponse {
  const limit = Number(context.args.limit ?? 20);
  const children = fs.readdirSync(context.project.path).sort((a, b) => a.localeCompare(b));
  for (const child of children.slice(0, limit)) {
    process.stdout.write(`${child}\n`);
  }
  return { success: true, message: `Listed up to ${limit} entries.` };
}

export function findProjectFileDirs(projectRoot: string, filename: string): string[] {
  const discovered: string[] = [];
  walkProjectTree(projectRoot, (currentDir, filenames) => {
    if (filenames.includes(filename)) {
      discovered.push(path.resolve(currentDir));
    }
  });
  return discovered.sort((left, right) => left.localeCompare(right));
}

function runCommandAcrossDirs(
  projectRoot: string,
  targetDirs: string[],
  command: string[],
  successLabel: string,
): BuiltinScriptResponse {
  const failures: string[] = [];
  for (const targetDir of targetDirs) {
    const relativeDir = path.relative(projectRoot, targetDir);
    const label = relativeDir === "" ? "." : relativeDir.split(path.sep).join("/");
    process.stdout.write(`Running in ${label}: ${command.join(" ")}\n`);
    const result = spawnSync(command[0], command.slice(1), {
      cwd: targetDir,
      encoding: "utf8",
      shell: false,
    });
    if (result.stdout?.trim()) {
      process.stdout.write(result.stdout.trimEnd() + "\n");
    }
    if (result.stderr?.trim()) {
      process.stderr.write(result.stderr.trimEnd() + "\n");
    }
    if (result.status !== 0) {
      failures.push(label);
      process.stdout.write(`Command failed in ${label} with exit code ${result.status ?? "unknown"}\n`);
    }
  }
  if (failures.length > 0) {
    return {
      success: false,
      message: `${successLabel} failed in ${failures.length} location(s): ${failures.join(", ")}`,
    };
  }
  return {
    success: true,
    message: `${successLabel} completed in ${targetDirs.length} location(s).`,
  };
}

function walkProjectTree(root: string, visitor: (currentDir: string, filenames: string[]) => void): void {
  const stack = [root];
  while (stack.length > 0) {
    const currentDir = stack.pop()!;
    const entries = fs.readdirSync(currentDir, { withFileTypes: true });
    const directories: string[] = [];
    const filenames: string[] = [];
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!RUNTIME_SKIP_DIRS.has(entry.name)) {
          directories.push(path.join(currentDir, entry.name));
        }
      } else if (entry.isFile()) {
        filenames.push(entry.name);
      }
    }
    visitor(currentDir, filenames);
    directories.sort((a, b) => a.localeCompare(b)).reverse().forEach((directory) => stack.push(directory));
  }
}

function resolveExecutable(command: string): string | null {
  const pathValue = process.env.PATH ?? "";
  const extensions = process.platform === "win32" ? [".cmd", ".exe", ".bat", ""] : [""];
  for (const directory of pathValue.split(path.delimiter)) {
    for (const extension of extensions) {
      const candidate = path.join(directory, `${command}${extension}`);
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }
  }
  return null;
}
